src/test/librbd/test_librbd.cc

```diff
@@ -2531,6 +2531,158 @@ TEST_F(TestLibRBD, TestCloneEncryption)
   rados_ioctx_destroy(ioctx);
 }
 
+TEST_F(TestLibRBD, LUKS1UnderLUKS2WithoutResize)
+{
+  REQUIRE_FEATURE(RBD_FEATURE_LAYERING);
+  REQUIRE(!is_feature_enabled(RBD_FEATURE_STRIPINGV2));
+  REQUIRE(!is_feature_enabled(RBD_FEATURE_JOURNALING));
+
+  librados::IoCtx ioctx;
+  ASSERT_EQ(0, _rados.ioctx_create(m_pool_name.c_str(), ioctx));
+
+  librbd::RBD rbd;
+  std::string parent_name = get_temp_image_name();
+  std::string clone_name = get_temp_image_name();
+  uint64_t data_size = 25 << 20;
+  uint64_t luks1_meta_size = 4 << 20;
+  uint64_t luks2_meta_size = 16 << 20;
+  std::string parent_passphrase = "parent passphrase";
+  std::string clone_passphrase = "clone passphrase";
+
+  {
+    int order = 22;
+    ASSERT_EQ(0, create_image_pp(rbd, ioctx, parent_name.c_str(),
+                                 luks1_meta_size + data_size, &order));
+    librbd::Image parent;
+    ASSERT_EQ(0, rbd.open(ioctx, parent, parent_name.c_str(), nullptr));
+
+    librbd::encryption_luks1_format_options_t fopts = {
+        RBD_ENCRYPTION_ALGORITHM_AES256, parent_passphrase};
+    ASSERT_EQ(0, parent.encryption_format(RBD_ENCRYPTION_FORMAT_LUKS1, &fopts,
+                                          sizeof(fopts)));
+
+    ceph::bufferlist bl;
+    bl.append(std::string(data_size, 'a'));
+    ASSERT_EQ(data_size, parent.write(0, data_size, bl));
+
+    ASSERT_EQ(0, parent.snap_create("snap"));
+    ASSERT_EQ(0, parent.snap_protect("snap"));
+    uint64_t features;
+    ASSERT_EQ(0, parent.features(&features));
+    ASSERT_EQ(0, rbd.clone(ioctx, parent_name.c_str(), "snap", ioctx,
+                           clone_name.c_str(), features, &order));
+  }
+
+  {
+    librbd::Image clone;
+    ASSERT_EQ(0, rbd.open(ioctx, clone, clone_name.c_str(), nullptr));
+
+    librbd::encryption_luks2_format_options_t fopts =
+        {RBD_ENCRYPTION_ALGORITHM_AES256, clone_passphrase};
+    ASSERT_EQ(0, clone.encryption_format(RBD_ENCRYPTION_FORMAT_LUKS2, &fopts,
+                                         sizeof(fopts)));
+
+    librbd::encryption_luks_format_options_t opts1 = {parent_passphrase};
+    librbd::encryption_luks_format_options_t opts2 = {clone_passphrase};
+    librbd::encryption_spec_t specs[] = {
+        {RBD_ENCRYPTION_FORMAT_LUKS, &opts2, sizeof(opts2)},
+        {RBD_ENCRYPTION_FORMAT_LUKS, &opts1, sizeof(opts1)}};
+    ASSERT_EQ(0, clone.encryption_load2(specs, std::size(specs)));
+
+    uint64_t size;
+    ASSERT_EQ(0, clone.size(&size));
+    EXPECT_EQ(data_size + luks1_meta_size - luks2_meta_size, size);
```

Comment:
Would it be the expected behaviour  for users, that after changing the cloned image format the data area may be not the same as the parent data area? Shouldn't we resize the clone image accordingly when formatting?

Reply:
> Would it be the expected behaviour for users, that after changing the cloned image format the data area may be not the same as the parent data area?

As much as I dislike that, that is the expected behavior.  If the crypto header in the parent is larger that the crypto header in the clone and the user didn't take steps to reserve the space in the clone prior to creating it, there is nothing we can do -- at least not with librbd API as is.

> Shouldn't we resize the clone image accordingly when formatting?

Two things:

- we can't resize the clone on the user's behalf because `rbd_clone` API is blissfully unaware of encryption
- even if we could resize the clone on the user's behalf, that wouldn't expose the missing part of the parent DATA area because parent overlap would remain at the original clone size

I see three ways out of this:

(A) Introduce some sort of "reset parent overlap" API.  Cloning for this case would look like:

- snap create/protect
- clone
- format the clone
- grow the clone to compensate for the difference in crypto header sizes
- reset overlap to the new size

(B) Require the user to reserve the space in the clone prior to creating it.  Cloning for this case would look like:

- grow the parent to reserve space for the difference in crypto header sizes
- snap create/protect
- clone
- format the clone
- shrink the parent back to the original size

(C) Rethink our approach to how encryption is exposed by librbd: introduce new higher-level encryption-aware APIs or at least make `rbd_encryption_format` optionally handle resizing and resetting parent overlap.  Formatting the clone for this case should become a single-step operation -- all details and image metadata manipulation should be encapsulated.  Another example is formatting a regular image -- it could be optionally resized to compensate for the added crypto header.

I went with (B) in this PR because (A) requires adding a new class method at the very least and (C) is an even longer shot.  My sense is that people would generally use either LUKS1 or LUKS2 and wouldn't mix them too often.

Reply:
Actually I don't think (A) and (B) makes sense, because when cloning we don't change the format (AFAIR it). It is only when we run `rbd_encryption_format` on a clone. So I was thinking about making `rbd_encryption_format` to do this automatically. I don't see a need to make as an option though.

> Another example is formatting a regular image: it could be optionally resized to compensate for the added crypto header -- this doesn't happen today.

Actually I think I like this idea, but again, not optionally but always. Making it optional does not differ much then just asking the user to do resize as additional step.

Reply:
> Actually I don't think (A) and (B) makes sense, because when cloning we don't change the format (AFAIR it). It is only when we run `rbd_encryption_format` on a clone. So I was thinking about making `rbd_encryption_format` to do this automatically. I don't see a need to make as an option though.

It would have clearer if I split (C) into two options:

(C-1) Fully encryption-aware API, where `rbd_create`-like and `rbd_clone`-like methods handle everything from creating an image/clone to formatting it and adjusting as necessary.  This could enable a rich set of "on-the-fly re-encryption" use cases, such as importing a plaintext image and turning it into an encrypted image at the same time or deep copying an image encrypted with one format into an image encrypted with another format.  `rbd_encryption_format` method would be removed.

(C-2) Make `rbd_encryption_format` method smarter to reduce the number steps it takes to create an encrypted image/clone.

It sounds like you prefer (C-2).  But note that (C-2) is would still need (A) on the backend -- since creating an image/clone and formatting it are separate steps we would still need a way to reset the parent overlap as part of formatting.

Reply:
Actually (C-1) looks appealing to me but (I my be wrong) I think we had a discussion with @orozery long time ago why we didn't go that way and there was some reason. So just for this reason I was thinking (C-2) would be more realistic.

Reply:
Yeah, (C-2) is definitely more realistic but still not trivial -- the implications of resetting parent overlap need to be carefully considered (apart from adding a new class method, etc).  And since changing `rbd_encryption_format` semantics would affect not just clones but also regular images (as you suggested making resizing unconditional), I consider it to be out of scope of this PR.

Reply:
... and since (B) doesn't bring any API modifications and just requires the user to temporarily grow the parent image, (C-2) can always be implemented on top if someone gets to it.

Reply:
> I went with (B) in this PR because (A) requires adding a new class method at the very least and (C) is an even longer shot. My sense is that people would generally use either LUKS1 or LUKS2 and wouldn't mix them too often.

I agree. Users should choose one format and stick to it. I don't think we should complicate ourselves too much to ease on users that will mix.
BTW, from a cloud provider perspective, the end-user is not aware of this whole crypto header sizes thing. It is the cloud provider's responsibility to handle all of the resizing. So even if the current approach is cumbersome, it is not user-facing (in the case of a cloud provider).

Reply:
So, who I understand it. With the current code, if a user have a luks1 base image, and expect that one may want to clone it and format to luks2, to avoid data loss, before creating a snapashot she have to expand the base image by luks2 - luks1 difference, and after creating the snapshot the image may be shrinked back.

I expect a lot of confusion here, i.e. (how a clone caller may be sure the parent image has already been expanded). On the other hand I agree that it is most likely very rare case and we may not need to spend a lot of effort on it. Personally, I would then just forbid formatting a clone if the new crypto header is bigger. But if you still think it should be allowed then at least clearly documenting this case would be nice.

Reply:
> So, how I understand it. With the current code, if a user have a luks1 base image, and expect that one may want to clone it and format to luks2, to avoid data loss, before creating a snapashot she have to expand the base image by luks2 - luks1 difference, and after creating the snapshot the image may be shrinked back.

Correct, and this also goes for creating an encrypted clone of a plaintext parent.  However I wouldn't call it data loss since the data stays there in the parent.  It's not lost, it's just that the last 4M to 16M of the parent just aren't available to the clone.

> I expect a lot of confusion here, i.e. (how a clone caller may be sure the parent image has already been expanded).

Expanding the parent multiple times (or by more than the required difference, e.g. using some fixed amount like 1G) is fine.

> On the other hand I agree that it is most likely very rare case and we may not need to spend a lot of effort on it. Personally, I would then just forbid formatting a clone if the new crypto header is bigger. But if you still think it should be allowed then at least clearly documenting this case would be nice.

I think it should be allowed because the whole thing would be too constrained otherwise.  Like Or said, in cases where there is some management layer sitting between the user and the actual images, these details are trivial to automate.

I tried to document this in the Examples section (https://ceph--48618.org.readthedocs.build/en/48618/rbd/rbd-encryption/#examples).  Feel free to suggest stroger wording!

Reply:
The doc (examples) looks clear to me. I just missed it somehow when reviewing. We might want to add a warning note in the man's `encryption format` command description about the effect when formatting a clone with a new header lager than parent. But I don't insist.